### ISSUE TITLE
EPIC: Unify and extend HTTP Client Configuration Capabilities

### ISSUE THREAD

**Issue by jakobmoellerdev** [open-component-model/ocm-project#664]
**Description**

We want to be able to globally (and locally) configure HTTP Client specific settings such as 

- PROXY Settings
- Certificate Settings
- Secure / Insecure Flags

as well as inject global handlers such as

- Logging Middleware
- Middleware for Tracing (down the line)
- Injection of DNS overrides for local testing

For this we need to establish a common configuration structure that can be used within OCM globally as well as be injected into accessmethods on demand later on (to locally override).

An example config might look like this:
https://gist.github.com/hilmarf/130eae7fad45e3602f1b061bbb6abf49#file-ocmconfig-L91-L106

**Comment by piotrjanik** [comment]
To assist with refining this Epic, the following requests have been made so far:
- [x] timeouts & retry policy - allows fine-tuning how a connection is established
  - requested in: https://github.com/open-component-model/ocm/issues/1731
  - draft: https://github.com/piotrjanik/open-component-model/blob/95cf36356d3651c046165cd510c08c982318bf20/bindings/go/configuration/http/v1alpha1/spec/config.go#L86
- [x] TLS:  
  - [x] allow insecure - disable TLS validation
  - [x] custom certificate:
    - [x] CA - allow providing custom CA cert in PEM format
    - [ ] Client - allow providing Client cert (_oras-go v2 supports mTLS_) - **_shouldn't that be part of Credentials_**❓ 
- [x] Proxy - allow setting an HTTP proxy
- [x] per-host config - this would enable adjusting settings per host, acting as an override layer
- [ ] Tracing - I think we should implement Otel standards here
- [ ] Logging - opt-in logging requests for debugging purposes (incl. secrets obfuscation)
- [ ] DNS override - allow overriding DNS resolution

Only Timeouts need to be implemented in legacy OCM; all other settings should be implemented in the new OCM.
For each item we consider valuable, we can create an issue, and the first one would also establish a common ground in the form of the HTTP config struct.

This is what an HTTP config could look like:

```yaml
type: http.config.ocm.software/v1alpha1


# --- Timeouts ---
timeout: "0s"
tcpDialTimeout: "30s"
tcpKeepAlive: "30s"
tlsHandshakeTimeout: "10s"
responseHeaderTimeout: "10s"
idleConnTimeout: "90s"

# --- Retry ---
retryMinWait: "200ms"
retryMaxWait: "3s"
retryMaxRetry: 5

# --- TLS / Certificate Settings (new) ---
tls:
  insecure: false                          # skip TLS certificate verification
  caCerts:                                 # additional trusted CA certificates
    - /etc/ssl/certs/corporate-ca.pem
    - /etc/ssl/certs/internal-ca.pem
# --- Proxy Settings (new) ---
proxy:
  httpProxy: "http://proxy.corp.example.com:8080"
  httpsProxy: "http://proxy.corp.example.com:8443"
  noProxy: "localhost,127.0.0.1,.corp.example.com,10.0.0.0/8"

# --- DNS Overrides (new, for local testing) ---
dnsOverrides:
  "registry.example.com": "127.0.0.1"
  "api.internal.corp": "192.168.1.100"

# --- Middleware / Handlers (new) ---
logging:
  enabled: true
  level: "debug"                           # debug, info, warn, error
  includeHeaders: false                    # log request/response headers
  includeBody: false                       # log request/response bodies (careful with large payloads)
tracing:
  enabled: false

# Per-host overrides. Each key is a hostname (or hostname:port).
# Any field set here overrides the corresponding global value for that host.
hosts:
  "ghcr.io:443":
    timeout: "60s"
    retryMaxRetry: 10
    tls:
      insecure: false
      caCerts:
        - /etc/ssl/certs/github-ca.pem
```

**Comment by fabianburth** [comment]
TLDR (discussion around request-scoped ocm configs);
- we initialize the plugin per reconcile loop with the currently provided config
- this means, we do not touch the binding level interfaces
- we do not cache plugins (initially)

(lets try to solve this WITHOUT immediately trying to solve the caching issue)

**Comment by fabianburth** [comment]
TLDR (cache discussion);
- we want to drop the component level cache of the controllers in favor of the blob cache
- we drop the asynchronous resolution (worker pool) from the controllers and move them into the library OR the blob cache and manifest resolution cache needs to be completely concurrency safe
- we did not really have a conclusion on how to do this refactoring - we kind of agreed that we would not really want to move the asynchronous resolution into the library as we lack the interfaces for async behaviour
- we'd accept that the worker pool schedules a worker EVEN for cached blobs BECAUSE the worker pool does not know about the cache state (so whether the underlying blob of the component is already cached)
- the problem here is that we need the `isCached` discriminator to determine whether we leave the reconcile loop or whether we can stay in the reconcile loop and immediately return the component (without the `isCached` discriminator, we would have to schedule a work item everytime, even if it's actually cached by the blob cache)

**Main Issues:**
- the oci layer (binding layer) does not know the "reference trust" (whether the referenced descriptor really has the digest in the component reference)
- to optimize the async behaviour, we'd need to move this into the bindings

**Questions**
- our `ocm get cv` theoretically also HAS TO support signature verification (and digest verification for `--recursive`)

**Idea:**
- **drop worker pool but increase amount of parallel workers**
  - verification service stays in controller
  - blob and descriptor lookup happens in oci
  - blob cache (path) is chosen based on config (credential) cache key
  - the cache is **authorized** BUT **untrusted** (trust comes from the verification service and we have a size limit on untrusted content) 
  - the cache can do this discrimination as it actually has to resolve and know the credentials that it would use to resolve the content
  - we need to evaluate whether this can achieve similar performance characteristics as the resolution service
